### PR TITLE
move and rename extract_component_subset

### PR DESCRIPTION
**Pull request by tjhei**
This does not change functionality but allows me to use the function in other .cc files later.

### CODE REVIEW

File: source/simulator/introspection.cc [hunk 1]
```diff
@@ -166,6 +166,90 @@ namespace aspect
...
+  {
+    std::vector<unsigned char> local_component_association (fe.dofs_per_cell,
+                                                            (unsigned char)(-1));
+
+    // compute the component each local dof belongs to.
+    // if the shape function is primitive, then this
+    // is simple and we can just associate it with
+    // what system_to_component_index gives us
+    for (unsigned int i=0; i<fe.dofs_per_cell; ++i)
+      if (fe.is_primitive(i))
```
Comment (bangerth) [inline source/simulator/introspection.cc]: are we ever using non-primitive elements? we could avoid a lot of complexity if we assume that

File: source/simulator/introspection.cc [hunk 2]
```diff
@@ -166,6 +166,90 @@ namespace aspect
   }
 
 
+  /**
+   * This is an internal deal.II function stolen from dof_tools.cc
+   */
+  template <int dim, int spacedim>
```
Comment (bangerth) [inline source/simulator/introspection.cc]: can you add the deal.II documentation of the function here?

File: source/simulator/introspection.cc [hunk 3]
```diff
@@ -166,6 +166,90 @@ namespace aspect
...
+            ==
+            local_component_association.end(),
+            ExcInternalError());
+
+    return local_component_association;
+  }
+
+  template <int dim>
+  IndexSet extract_component_subset(const DoFHandler<dim> &dof_handler,
+                                    const ComponentMask &component_mask)
```
Comment (bangerth) [inline source/simulator/introspection.cc]: I don't like the name. Can we call it `extract_dofs_for_component_subset()` or something?

**Comment by tjhei** [comment]
addressed your comments

**Comment by bangerth** [comment]
OK.

**Pull request merged**